### PR TITLE
[SPARK-22981][SQL] Fix incorrect results of Casting Struct to String

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr fixed the issue when casting structs into strings;
```
scala> val df = Seq(((1, "a"), 0), ((2, "b"), 0)).toDF("a", "b")
scala> df.write.saveAsTable("t")                                        
scala> sql("SELECT CAST(a AS STRING) FROM t").show
+-------------------+
|                  a|
+-------------------+
|[0,1,1800000001,61]|
|[0,2,1800000001,62]|
+-------------------+
```
This pr modified the result into;
```
+------+
|     a|
+------+
|[1, a]|
|[2, b]|
+------+
```

## How was this patch tested?
Added tests in `CastSuite`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -259,6 +259,29 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
         builder.append("]")
         builder.build()
       })
+    case StructType(fields) =>
+      buildCast[InternalRow](_, row => {
+        val builder = new UTF8StringBuilder
+        builder.append("[")
+        if (row.numFields > 0) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Probably, it seems we have no chance to hit `row.numFields == 0` here though, I just leave this for strict checks.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
@cloud-fan ping

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: do we need this? even `row.numFields == 0`, the following code is a no-op

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -732,6 +755,39 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
...
 
+  private def writeStructToStringBuilder(
+      st: Seq[DataType],
+      row: String,
+      buffer: String,
+      ctx: CodegenContext): String = {
+    val structToStringCode = st.zipWithIndex.map { case (ft, i) =>
+      val fieldToStringCode = castToStringCode(ft, ctx)
+      val funcName = ctx.freshName("fieldToString")
+      ctx.addNewFunction(funcName,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: no need to create a function, it's called only once.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Actually we can create functions, for data types that appeared more than once among struct fields.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: BTW here we may hit 64kb compile error if there are a lot of fields in this struct. We should use `ctx.splitExpressionsWithCurrentInput`

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: ok, l'll update soon

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, merging to master/2.3!

**Comment by maropu** [comment]
Thanks! Next, I'll fix `showString` though, one question; currently casting binary to string is different between [Cast](https://github.com/apache/spark/blob/2250cb75b99d257e698fe5418a51d8cddb4d5104/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L205) and [showString](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L254). Which one is expected? 

FYI: This casting behaviour is different between mysql and postgresql;
```
postgres=# create table t (a bytea);
CREATE TABLE
postgres=# insert into t values('abc');
INSERT 0 1
postgres=# SELECT CAST(a AS TEXT) from t;
    a     
----------
 \x616263

mysql> create table t(a blob);
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values('abc');
Query OK, 1 row affected (0.00 sec)

mysql> select CAST(a AS char) From t;
+-----------------+
| CAST(a AS char) |
+-----------------+
| abc             |
+-----------------+
1 row in set (0.00 sec)
```

**Comment by maropu** [comment]
hive one is the same with mysql one;
```
hive> create table t(a BINARY);
OK
hive> INSERT INTO t values('abc');
OK
hive> select CAST(a AS STRING) from t;
OK
abc
```

**Comment by cloud-fan** [comment]
show binary as string and cast binary to string seems different to me, let's stick with what it is. BTW it's pretty dangerous to change the behavior of cast to be different with Hive.

**Comment by maropu** [comment]
ok, I'll make a follow-up for `showString` later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 3]
```diff
@@ -259,6 +259,29 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
         builder.append("]")
         builder.build()
       })
+    case StructType(fields) =>
+      buildCast[InternalRow](_, row => {
+        val builder = new UTF8StringBuilder
+        builder.append("[")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:265]: Are there any reasons for using the same brackets `[]` for Struct, Map and Array?

How about `[]` for arrays, `{}` for structs and `<>` for maps or somehow distinguish them. At least, it is not consistent with `toHiveString` which prints `{}` for structs and maps.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:265]: I think we have no strong reason. In this PR, I just followed the existing format. Looks like we can change it to `{` for consistency. cc: @cloud-fan

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:265]: I'm OK with the new format.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:265]: Here is the PR https://github.com/apache/spark/pull/29308 . Please, review it.

**Pull request closed without merging**